lib/appmap/event.rb

```diff
@@ -60,16 +60,37 @@ def display_string(value)
           final ? value_string : encode_display_string(value_string)
         end
 
-        def object_properties(hash_like)
-          hash = hash_like.to_h
-          hash.keys.map do |key|
-            {
+        def add_schema(param, value, always: false)
+          return unless always || AppMap.parameter_schema?
+
+          if value.blank? || value.is_a?(String)
+            # pass
+          elsif value.is_a?(Enumerable)
+            if value.is_a?(Hash)
+              param[:properties] = object_properties(value)
+            elsif value.respond_to?(:first) && value.first
+              param[:properties] = object_properties(value.first)
+            end
+          elsif value.respond_to?(:as_json)
+            add_schema param, JSON.parse(value.to_json), always: always
```

Comment:
Why not use `value.as_json` directly instead of doing a potentially expensive roundtrip?

Reply:
Some objects that implement `to_json` don't implement `as_json`. `as_json` is an ActiveModel thing - so it'll be common, but not required. Still, I could try `as_json` first and then fall back on the round-trip.

Reply:
But you're looking for `as_json` in the condition, so all objects will have it at this point. Unless that's not what you wanted :)